Add improved support for mobile platforms

vaidas-unity opened the pull request:
**Purpose of this PR**:
This PR extends the functionality of the BuildReportInspector package by adding improved support for Android and iOS platforms:
* Display the contents of the .apk/.oob/.aab (Android) or .ipa (iOS).
* Display the estimated application download size from its corresponding store to the end-user.
* Display the architectures available in the build.
* Embed a simple test project into the repository.
* Fix support for deprecated .NET 3.5.
* Omit the need for analyzing the serialized BuildReport asset.


**Release Notes**:
Add improved support for iOS and Android.
Fix #1 - incomplete Assets list shown when there are many entries.
Fix .NET 3.5 support.

**Testing status**:
Manual testing - in progress (internal QA, ask @vaidas-unity for document link).
Automated tests - TBA.

**Overall Product Risks**:
Technical Risk: 2
Halo Effect: 2

vincent-unity commented on com.unity.build-report-inspector/Editor/BuildReportInspector/Mobile/PostBuildSetup.cs:
```diff
@@ -0,0 +1,49 @@
+        private const string k_GuidFileName = "UnityBuildGuid.txt";
+        private static string s_LastBuildGuid;
+        
+        public int callbackOrder { get; }
```
You rely on default 0 on purpose?

vincent-unity commented on com.unity.build-report-inspector/Editor/BuildReportInspector/Mobile/PostBuildSetup.cs:
```diff
@@ -0,0 +1,49 @@
+                return;
+                    
+            // Save the guid for BuildPostProcess callback.
+            s_LastBuildGuid = report.summary.guid.ToString();
```
Move this after the `if` statement below?

vaidas-unity replied:
This is not needed for iOS, that's why I kept it above the if

vincent-unity commented on com.unity.build-report-inspector/Editor/BuildReportInspector/Mobile/Tools/Utilities.cs:
```diff
@@ -0,0 +1,36 @@
+using System.Diagnostics;
+using System.IO;
+
+namespace Unity.Mobile.BuildReport.Tools
```
This namespace is not consistent to others. On purpose?

vaidas-unity replied:
Will fix

vaidas-unity replied:
Fixed

vincent-unity commented on com.unity.build-report-inspector/Editor/BuildReportInspector/Mobile/Tools/Utilities.cs:
```diff
@@ -0,0 +1,36 @@
+#if UNITY_EDITOR
```
No need for #if as this script is under `Editor` folder.

vaidas-unity replied:
Fixed

vincent-unity commented on com.unity.build-report-inspector/Editor/BuildReportInspector/Mobile/MobileHelper.cs:
```diff
@@ -0,0 +1,84 @@
+    internal static class MobileHelper
+    {
+        internal static IPlatformUtilities s_PlatformUtilities;
+        internal static string AppendixSavePath => Path.Combine(Directory.GetParent(Application.dataPath).FullName, "Library/MobileReports");
```
Again, please do check on 2018.x with .net 3.5 scripting backend.

vaidas-unity replied:
Fixed

vincent-unity commented on com.unity.build-report-inspector/Editor/BuildReportInspector/Mobile/MobileHelper.cs:
```diff
@@ -0,0 +1,84 @@
+            if (!Directory.Exists(AppendixSavePath))
+                Directory.CreateDirectory(AppendixSavePath);
+
+            if (s_PlatformUtilities != null)
```
Curious about what will happen if you switch between iOS and Android platforms.

vaidas-unity replied:
Nothing, the s_PlatformUtilities is lost upon domain reload and re-initialized.

vincent-unity replied:
Ah, yes, forgot this..

vincent-unity commented on com.unity.build-report-inspector/Editor/BuildReportInspector/Mobile/Apple/AppleUtilities.cs:
```diff
@@ -0,0 +1,120 @@
+                
+                if (appleArchInfos.Count < 1)
+                {
+                    throw new Exception($"Couldn't extract architecture info from application {applicationPath}");
```
As I remembered, $ won't work on 3.5 scripting backend.

vaidas-unity replied:
Fixed

vaidas-unity commented on com.unity.build-report-inspector/Editor/BuildReportInspector/Mobile/Android/AndroidUtilities.cs:
```diff
@@ -0,0 +1,245 @@
+            Aab
+        }
+
+        private static bool IsTestEnvironment => !string.IsNullOrEmpty(Environment.GetEnvironmentVariable("BOKKEN_RESOURCEID"));
```
Fixed

vaidas-unity commented on com.unity.build-report-inspector/Editor/BuildReportInspector/Mobile/Apple/AppleUtilities.cs:
```diff
@@ -0,0 +1,120 @@
+using System.Collections.Generic;
+using UnityEditor;
+
+namespace Unity.BuildReportInspector.Mobile.Apple
```
Fixed

vaidas-unity commented on com.unity.build-report-inspector/Editor/BuildReportInspector/Mobile/Android/AndroidUtilities.cs:
```diff
@@ -0,0 +1,245 @@
+#if UNITY_ANDROID
```
I agree, will fix

vaidas-unity replied:
Fixed

vaidas-unity commented on com.unity.build-report-inspector/Editor/BuildReportInspector/Mobile/Apple/AppleUtilities.cs:
```diff
@@ -0,0 +1,120 @@
+using System;
```
Fixed

vincent-unity commented on TestProject/Packages/manifest.json:
```diff
@@ -0,0 +1,10 @@
+  "disableProjectUpdate": true,
+  "dependencies": {
+    "com.unity.build-report-inspector": "file:../../com.unity.build-report-inspector",
+    "com.unity.ide.rider": "2.0.2"
```
Since you added rider, probably would be good to add visualstudio, vs code as well?

vaidas-unity replied:
I added this by accident, will remove 🙂

vincent-unity commented on com.unity.build-report-inspector/Editor/Mobile/AndroidUtilities.cs:
```diff
@@ -0,0 +1,283 @@
+        {
+            get
+            {
+#if UNITY_2018
```
I remembered you are going to make this package only compatible with 2019.3 for mobile in yesterday's weekly meeting, correct me if I'm wrong. Then probably we can remove all these ifdef in the future.

vaidas-unity replied:
Fixed

Wilfrid-Unity commented on com.unity.build-report-inspector/Editor/csc.rsp:
```diff
@@ -0,0 +1,2 @@
+-r:System.IO.Compression.FileSystem.dll
```
Is it necessary to have these **csc.rsp** files in the package?

vaidas-unity replied:
Yes, otherwise it is not possible to use the C# ZipFile class, which is used for displaying apk/app contents.

Wilfrid-Unity commented on com.unity.build-report-inspector/README.md:
```diff
@@ -5,20 +5,17 @@ This package contains an Editor script which implements an inspector for the Bui
 
 This package is provided as-is, with no support from Unity Technologies. We plan to add a built-in and supported UI for the BuildReport feature in a future version of Unity, but until then, this package serves as a demonstration on how you can access the BuildReport information today.
 
-In particular, this package gets the information it can from the BuildReport Scripting API (https://docs.unity3d.com/ScriptReference/Build.Reporting.BuildReport.html), but some information in the BuildReport object is not yet exposed through public APIs.  
```
Did you intend to change these sections?
It seems the changes were not reflected in the README.md file under the repository root folder?

vaidas-unity replied:
Yes, I shouldn't have changed this, will fix

vaidas-unity replied:
Fixed - synced all of the md files. Also removed the section about serializedFile because the serialized file will no longer be analyzed on 2019.3 onwards, as the PackedAssets API has been exposed.

Wilfrid-Unity commented on com.unity.build-report-inspector/README.md:
```diff
@@ -50,6 +47,34 @@ For platforms which support engine code stripping, a list of all engine modules
+When BuildOptions.DetailedBuildReport is passed to [BuildPipeline.BuildPlayer](https://docs.unity3d.com/ScriptReference/BuildPipeline.BuildPlayer.html), a list describing which scenes are using each asset of the build, is provided in the BuildReport.
 
 <img src="Documentation~/images/ScenesUsingAssets.png" width="400">
+
```
I think you need to update file com.unity.build-report-inspector/Documentation~/com.unity.build-report-inspector.md as well, it is the source for online package documentation page https://docs.unity3d.com/Packages/com.unity.build-report-inspector@0.1/manual/index.html

vaidas-unity replied:
Mirrored the same info to com.unity.build-report-inspector.md, thanks

Wilfrid-Unity commented on .yamato/promotion.yml:
```diff
@@ -31,17 +32,6 @@ promotion_test_{{ platform.name }}_{{ editor.version }}:
 
 promotion_test_trigger:
   name: Promotion Tests Trigger
-  agent:
```
It is not required to define an agent? 
Do the Package Promotion tests still work without these lines?

vaidas-unity replied:
Because this job doesn't execute any commands, it doesn't need an agent. More info on Virtual Jobs here: https://internaldocs.hq.unity3d.com/yamato/yaml/job-definitions/#virtual-jobs

Wilfrid-Unity replied:
Thanks for the link!

vaidas-unity commented:
> Thanks Vaidas for adding these features.
> 
> I only checked that there were no regressions regarding the older features, but I had an issue after trying the following steps:
> 
> * I created a new project with Editor version 2020.2.0a7.826 (trunk revision cea774d83c47), that did not have mobile support installed
> * I placed a cube with a custom unlit shader, and built the scene using the script from https://docs.unity3d.com/2020.1/Documentation/ScriptReference/BuildOptions.DetailedBuildReport.html
> * I added to the project BuildReportInspector revision [c5164bb](https://github.com/Unity-Technologies/BuildReportInspector/commit/c5164bb324bef75b57a3fc8ee4f5abc965139f81) (I installed it from my disk, where I had checked out the revision)
> * I clicked Window/Open Last Build Report
> * I got the following error displayed 999+ times in the console:
> 
> > ```
> > The appendix file does not exist at C:\unity-projects-current\wil-projects\check-build-report-mobile-pr\Assets/BuildReports/Mobile\b9761b87b7495474dbd4e42ef065a9cd.
> > UnityEngine.Debug:LogError (object)
> > Unity.BuildReportInspector.Mobile.MobileHelper:LoadMobileAppendix (string) (at C:/source-code/from-github-unity/BuildReportInspector/com.unity.build-report-inspector/Editor/Mobile/MobileHelper.cs:85)
> > Unity.BuildReportInspector.BuildReportInspector:get_mobileAppendix () (at C:/source-code/from-github-unity/BuildReportInspector/com.unity.build-report-inspector/Editor/BuildReportInspector.cs:47)
> > Unity.BuildReportInspector.BuildReportInspector:OnInspectorGUI () (at C:/source-code/from-github-unity/BuildReportInspector/com.unity.build-report-inspector/Editor/BuildReportInspector.cs:176)
> > UnityEngine.GUIUtility:ProcessEvent (int,intptr,bool&) (at C:/unity-pr2/Modules/IMGUI/GUIUtility.cs:189)
> > ```
> 
> Can you please add something to avoid such issues?
> 
> Thanks a lot also for having added tests, that's very helpful.

Fixed this issue, thanks for taking your time to test it.

Wilfrid-Unity commented:
> > Thanks Vaidas for adding these features.
> > I only checked that there were no regressions regarding the older features, but I had an issue after trying the following steps:
> > 
> > * I created a new project with Editor version 2020.2.0a7.826 (trunk revision cea774d83c47), that did not have mobile support installed
> > * I placed a cube with a custom unlit shader, and built the scene using the script from https://docs.unity3d.com/2020.1/Documentation/ScriptReference/BuildOptions.DetailedBuildReport.html
> > * I added to the project BuildReportInspector revision [c5164bb](https://github.com/Unity-Technologies/BuildReportInspector/commit/c5164bb324bef75b57a3fc8ee4f5abc965139f81) (I installed it from my disk, where I had checked out the revision)
> > * I clicked Window/Open Last Build Report
> > * I got the following error displayed 999+ times in the console:
> > 
> > > ```
> > > The appendix file does not exist at C:\unity-projects-current\wil-projects\check-build-report-mobile-pr\Assets/BuildReports/Mobile\b9761b87b7495474dbd4e42ef065a9cd.
> > > UnityEngine.Debug:LogError (object)
> > > Unity.BuildReportInspector.Mobile.MobileHelper:LoadMobileAppendix (string) (at C:/source-code/from-github-unity/BuildReportInspector/com.unity.build-report-inspector/Editor/Mobile/MobileHelper.cs:85)
> > > Unity.BuildReportInspector.BuildReportInspector:get_mobileAppendix () (at C:/source-code/from-github-unity/BuildReportInspector/com.unity.build-report-inspector/Editor/BuildReportInspector.cs:47)
> > > Unity.BuildReportInspector.BuildReportInspector:OnInspectorGUI () (at C:/source-code/from-github-unity/BuildReportInspector/com.unity.build-report-inspector/Editor/BuildReportInspector.cs:176)
> > > UnityEngine.GUIUtility:ProcessEvent (int,intptr,bool&) (at C:/unity-pr2/Modules/IMGUI/GUIUtility.cs:189)
> > > ```
> > 
> > 
> > Can you please add something to avoid such issues?
> > Thanks a lot also for having added tests, that's very helpful.
> 
> Fixed this issue, thanks for taking your time to test it.

Thanks for all the fixes, to me the pull request looks good to merge :)

vaidas-unity commented:
Merging. Will request additional QA testing on master branch.
